server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java

```diff
@@ -0,0 +1,258 @@
+         scheduledTimer = (timeout > 0) ? handler.getScheduler().schedule(() -> {
+            cache.removeListenerAsync(this);
+            synchronizer.resultFuture.complete(null);
+         }, timeout, TimeUnit.MILLISECONDS) : null;
+      }
+
+      private void deleteTimer() {
+         if (scheduledTimer != null)
+            scheduledTimer.cancel(true);
+         scheduledTimer = null;
+      }
+
+      @CacheEntryCreated
+      @CacheEntryModified
+      public void onEvent(CacheEntryEvent<Object, Object> entryEvent) {
+         try {
+            if (entryEvent.getValue() instanceof ListBucket) {
+               byte[] key = unwrapKey(entryEvent.getKey());
+               synchronizer.onEvent(key);
+            }
+         } catch (Exception ex) {
+            synchronizer.resultFuture.completeExceptionally(ex);
+         }
+      }
+
+      private byte[] unwrapKey(Object key) {
+         return key instanceof WrappedByteArray
+               ? ((WrappedByteArray) key).getBytes()
+               : (byte[]) key;
+      }
+   }
+
+   /**
+    * PollListenerSynchronizer
+    *
+    * This class synchronizes the access to a CompletableFuture `resultFuture` so
+    * that its final value will be completed either
+    * - with value v by an onListenerAdded() call if a not null value is found
+    * - otherwise by onEvent(k) call if the referred entry is not null or empty
+    *
+    */
+   public static class PollListenerSynchronizer {
+      private final ConcurrentLinkedQueue<Object> keyQueue;
+      private final CompletableFuture<Collection<byte[]>> resultFuture;
+      private final EmbeddedMultimapListCache<byte[], byte[]> multimapList;
+      private final List<byte[]> keys;
+      private final BiConsumer<? super Collection<byte[]>, ? super Throwable> whenCompleteConsumer;
+      private boolean canPollJustEventKey;
+
+      private PollListenerSynchronizer(List<byte[]> keys, EmbeddedMultimapListCache<byte[], byte[]> multimapList) {
+         keyQueue = new ConcurrentLinkedQueue<Object>();
+         resultFuture = new CompletableFuture<Collection<byte[]>>();
+         this.multimapList = multimapList;
+         this.keys = keys;
+         whenCompleteConsumer = (v, t) -> {
+            if (t != null) {
+               resultFuture.completeExceptionally(t);
+            } else if (v != null && !v.isEmpty()) {
+               resultFuture.complete(v);
+            } else {
+               if (keyQueue.poll() == this) {
+                  canPollJustEventKey = true;
+               }
+               if (!keyQueue.isEmpty()) {
+                  runPoll();
+               }
+            }
+         };
+      }
+
+      private void runPoll() {
+         if (canPollJustEventKey) {
+            var key = (byte[]) keyQueue.peek();
+            BLPOP.pollKeyValue(multimapList, key).whenComplete(whenCompleteConsumer);
+         } else {
+            BLPOP.pollAllKeys(multimapList, keys).whenComplete(whenCompleteConsumer);
+         }
+      }
```

Comment:
```suggestion
      private void runPoll(Object key) {
         if (canPollJustEventKey && key != this) {
            BLPOP.pollKeyValue(multimapList, (byte[]) key).whenComplete(whenCompleteConsumer);
         } else {
            BLPOP.pollAllKeys(multimapList, keys).whenComplete(whenCompleteConsumer);
         }
      }
```
This method can be refactored to pass in the key as all callers have access to it and it would be a lot safer imo as then this method doesn't need to access any variables that are protected by synchronized block.

Note we still have to pay attention to canPollJustEventKey though.